Lib/subprocess.py

```diff
@@ -809,6 +822,40 @@ def __init__(self, args, bufsize=-1, executable=None,
             else:
                 line_buffering = False
 
+        gids = []
+        if groups is not None:
+            if isinstance(groups, str):
+                raise ValueError("Groups must be a list, not a string")
+
+            for group in groups:
+                if isinstance(group, str):
+                    gids.append(grp.getgrnam(group).gr_gid)
+                elif isinstance(group, int):
+                    gids.append(group)
+                else:
+                    raise TypeError("Group must be a string or integer")
+
+            if not gids:
+                raise ValueError("Groups cannot be empty")
+
+            # make sure that the gids are all positive here so we can do less
+            # checking in the C code
+            for gid in gids:
+                if gid < 0:
```

Comment:
can we feasibly check it against a maximum gid_t value?

Reply:
The only way I have managed to find (from https://stackoverflow.com/questions/13795758/what-is-sys-maxint-in-python-3/22948228 ) would be
```python
import struct
...
if gid > 2 ** (struct.Struct('i').size * 8 - 1) - 1:
    raise ValueError(f"Group ID is too large, got {gid}")
```

Reply:
okay lets not do that...  double checking and looking at Objects/longobject.c it looks like PyLong_AsUnsignedLong() does bounds checking for us already for values that won't fit rather than aliasing to a truncated value.  yay.

Reply:
Is this the sort of hack we want in the standard library? It will bomb out eventually in the C code if it's outside MAX_INT